x-pack/plugins/apm/public/services/rest/callApi.ts

```diff
@@ -26,10 +26,38 @@ function fetchOptionsWithDebug(fetchOptions: KFetchOptions) {
   };
 }
 
+const cache = new Map();
+
+export function _clearCache() {
+  cache.clear();
+}
+
 export async function callApi<T = void>(
   fetchOptions: KFetchOptions,
   options?: KFetchKibanaOptions
 ): Promise<T> {
+  const cacheKey = JSON.stringify(fetchOptions);
```

Comment:
So the main problem I can see here is that the key will be dependent on the order the keys were added to the object, so two identical calls could result in a cache miss just because the options objects were constructed differently.

I think in general we won't run into this, but the day we are confused about why something we _expect_ to be cached isn't being cached, this will definitely be the reason why :P

Reply:
Ahh, good point. I actually did think about sorting the object, but that was more to avoid cache invalidation for the same input (previously the spec didn't guarantee order of keys but it does now).

I see two options: 
 - sort object before stringifying
 - using something like [object-hash](https://github.com/puleos/object-hash) to get a predictable, unique hash for the object

Reply:
This might do the trick:

```ts
function getCacheKey(unordered: FetchOptions) {
  const sortedKeys = Object.keys(unordered).sort() as Array<keyof FetchOptions>;
  const ordered = sortedKeys.reduce<FetchOptions>((acc, key) => {
    acc[key] = unordered[key];
    return acc;
  }, {});

  return JSON.stringify(ordered);
}
```

Reply:
Cool, you maybe don't even need the reduce if you want to just simply have ordered keys and values:
```ts
function getCacheKey(options: FetchOptions) {
  const sortedKeys = Object.keys(options).sort() as Array<keyof FetchOptions>;
  const sortedValues = sortedKeys.map(v => options[v]) // not sure about typing this
  return JSON.stringify([sortedKeys, sortedValues])
}
```

Reply:
Great, I'll do that 👍

Reply:
Bahh... I forgot: we have to do this recursively for deeply nested objects. :(

```ts
await callApi({
  pathname: `/api/kibana`,
  query: { end: '2010', start: '2009' }
});
```
vs
```ts
await callApi({
  query: { start: '2009', end: '2010' },
  pathname: `/api/kibana`
});
```

That rules out the simple solutions. I _could_ make it recursive but I feel I'm going into territory where it might be better to use something that already exists, eg. [object-hash](https://github.com/puleos/object-hash)

Reply:
Good point. object-hash is cool with me, but @joshdover also pointed out that Kibana already ships with this fwiw: https://github.com/substack/json-stable-stringify

Reply:
I feel it's overkill to sort the object, it's not very likely to cause cache misses within the same deployment. Even if it does, it's OK since it will be cached in subsequent calls anyways.

Reply:
@ogupte Do you also think using `object-hash` is overkill? I've added that as part of `getCacheKey`. If @jasonrhodes is fine with it I can revert back to using `JSON.stringify`.

Reply:
Commit where object-hash is added https://github.com/elastic/kibana/pull/33392/commits/17c5f1bc05542ff61c05b560fbfe1bd6e345efe9

Reply:
> @ogupte Do you also think using `object-hash` is overkill? I've added that as part of `getCacheKey`. If @jasonrhodes is fine with it I can revert back to using `JSON.stringify`.

I still feel it's overkill. JSON.stringify is still unlikely to result in a cache miss unless the implementation of object creation changes between API calls. It's ok if we want to guard against this, it's just my opinion that it might be premature at this point.

Reply:
Fair comment. I vote to keep `object-hash` and I don't think it's overkill for two reasons, tell me if I'm misunderstanding either:

a) The `JSON.stringify` behavior is unexpected. If we make the _n_ calls to the same endpoint in _n_ places, we'd expect the 2nd-n calls to be cached, but it won't work unless the objects are exactly constructed in the same way, and we'll forget that's why and spend time debugging the extra call(s). 

b) This cache object is basically a self-made memory leak, and if a user keeps Kibana APM open indefinitely, it'll grow forever. At least with the hash, we minimize the problem as much as possible.

Reply:
I'm mostly in favor of "Principle of least surprise", and I agree it will be surprising if two different methods don't use the same cache due to the order of properties. That being said, it is highly unlikely to happen, since I can't imagine why we would have to identical methods for calling the backend.

Good point about the memory leak. I've previously used [`lru-cache`](https://github.com/isaacs/node-lru-cache) to solve this. It's already part of Kibana so is "free" to pull in.

Reply:
Added LRU cache: https://github.com/elastic/kibana/pull/33392/commits/4b1e4003282d61d0ba4720c1b01ac8929825a4b5